include/nccl_ofi.h

```diff
@@ -396,8 +396,8 @@ struct nccl_net_ofi_send_comm {
 	 * @return	0 on success
 	 *		non-zero on error
 	 */
-	int (*regMr)(nccl_net_ofi_send_comm_t *send_comm, void *data, size_t size, int type,
-			      void **mhandle);
+	int (*regMr)(nccl_net_ofi_send_comm_t *send_comm, const nccl_ofi_mr_input_t *in, int type,
```

Comment:
I don't really understand the gain in this?  It seems like overkill?

Reply:
The call stack from here to fi_mr_regattr is: 

* 0. regMr
* 1. reg_mr_send_comm + reg_mr_recv_comm
* 2. reg_mr_ep
* 3. reg_mr_on_device
* 4. set_mr_req_attr
* 5. fi_mr_regattr

And on the sendrecv side it's:

* 0. regMr
* 1. reg_mr_send_comm + reg_mr_recv_comm
* 2. reg_mr_base_comm
* 3. reg_mr_base
* 4. register_mr_buffers
* 5. fi_mr_regattr

As you've rightly pointed out, most of the functions in the middle do need to care whatsoever if there is a dmabuf descriptor involved, they just need to be able to do MR cache lookups and they need to forward it through.  

Here are your options:

1. copy-paste all the code and create `_dmabuf` functions, adding fd and offset arguments which are always valid.

For RDMA, that's

* 0. regMrDmaBuf
* 1. reg_mr_dmabuf_send_comm + reg_mr_dmabuf_send_comm
* 2. reg_mr_dmabuf_ep
* 3. reg_mr_dmabuf_on_device
* 4. set_mr_req_attr_dmabuf
* 5. fi_mr_regattr

And on the sendrecv side it's:

* 0. regMrDmaBuf
* 1. reg_mr_dmabuf_send_comm + reg_mr_dmabuf_send_comm
* 2. reg_mr_dmabuf_base_comm
* 3. reg_mr_dmabuf_base
* 4. register_mr_buffers_dmabuf
* 5. fi_mr_regattr

That's a lot of duplicated code.

2. keep a single set of "middle" functions, but add fd and offset arguments, which are potentially invalid, where there's no mechanism to ensure that dmabuf base address calculations are being done consistently in all cases, and dma-buf specific code is a check for `(fd == -1)`, where nothing prevents someone from using `base + len` directly without first checking for (fd == -1) and also considering the offset. Bugprone and makes some of these functions take an absurd amount of arguments (eg: sendrecv's `register_mr_buffers` would be taking ten arguments). 

3. What I've done: consolidate the MR arguments from NCCL into an opaque type, define an interface that allows you to check the type and allows you to get the base address and offset of the MR input request in a consistent way, and then replace existing addr + len arguments with that opaque type.

Reply:
I'd duplicate the code and clean it up when we fix the object system with C++.  I would not assume all the code is always the same all the way through.

Reply:
I really just do not see where you're coming from on this whatsoever. I'd want a similar trivially constructable type to abstract over the MR inputs in C++, too.

> I would not assume all the code is always the same all the way through.

who is assuming this? any function which wants to react to the presence of dmabuf arguments or lack thereof simply checks `in->type` against `NCCL_OFI_MR_INPUT_IOVEC` or `NCCL_OFI_MR_INPUT_DMABUF`.

Reply:
The entire stacking of functions is because of the need to split send/recv communicators as different types, because our type system doesn't have multiple inheritance.  That's not something that should be a problem with C++.  But at this point, I just don't see the point in all these changes; the old prototype you decided to delete woudl have worked fine.

Reply:
How would the old interface have worked fine? Once we're in regMrDmaBuf we still have to call into every function below it and we still have the problem I'm raising above of needing to either duplicate 12 different functions (only two of which need to care about the type of the MR), or we need to add two new function arguments to 12 existing functions.

Why make this so complicated?